drivers/timer/sam0_rtc_timer.c

```diff
@@ -0,0 +1,290 @@
+{
+	while (RTC0->STATUS.reg & RTC_STATUS_SYNCBUSY) {
+		/* Wait for bus synchronization... */
+	}
+}
+
+#ifdef CONFIG_TICKLESS_KERNEL
+
+/*
+ * Reads RTC COUNT register. First a read request must be written to READREQ,
+ * then - when bus synchronization completes - the COUNT register is read and
+ * returned.
+ */
+static u32_t rtc_count(void)
+{
+	/* Request synchronization and read the COUNT register. */
+	RTC0->READREQ.reg = RTC_READREQ_RREQ;
+	rtc_sync();
+	return RTC0->COUNT.reg;
+}
+
+#endif /* CONFIG_TICKLESS_KERNEL */
+
+/*
+ * Initializes the RTC peripheral.
+ */
+int z_clock_driver_init(struct device *device)
+{
+	ARG_UNUSED(device);
+
+	/* Set up bus clock and GCLK generator. */
+	PM->APBAMASK.reg |= PM_APBAMASK_RTC;
+	GCLK->CLKCTRL.reg = GCLK_CLKCTRL_ID(RTC_GCLK_ID) | GCLK_CLKCTRL_CLKEN
+			    | GCLK_GEN(DT_RTC_SAM0_CLOCK_GENERATOR);
+
+	while (GCLK->STATUS.bit.SYNCBUSY) {
+		/* Synchronize GCLK. */
+	}
+
+	/* Reset module to hardware defaults. */
+	rtc_reset();
+
+	rtc_last = 0;
+
+	/* Configure RTC with 32-bit mode, configured prescaler and MATCHCLR. */
+	u16_t ctrl = RTC_MODE0_CTRL_MODE(0) | RTC_MODE0_CTRL_PRESCALER(0);
+#ifndef CONFIG_TICKLESS_KERNEL
+	ctrl |= RTC_MODE0_CTRL_MATCHCLR;
+#endif
+	rtc_sync();
+	RTC0->CTRL.reg = ctrl;
+
+	/* Enable RTC interrupt. */
+	NVIC_ClearPendingIRQ(DT_RTC_SAM0_IRQ);
+	IRQ_CONNECT(DT_RTC_SAM0_IRQ, DT_RTC_SAM0_IRQ_PRIORITY, rtc_isr, 0, 0);
+	irq_enable(DT_RTC_SAM0_IRQ);
+
+#ifdef CONFIG_TICKLESS_KERNEL
+	/* Tickless kernel lets RTC count continually and ignores overflows. */
+	RTC0->INTENSET.reg = RTC_MODE0_INTENSET_CMP0;
+#else
+	/* Non-tickless mode uses comparator together with MATCHCLR. */
+	rtc_sync();
+	RTC0->COMP[0].reg = CYCLES_PER_TICK;
+	RTC0->INTENSET.reg = RTC_MODE0_INTENSET_OVF;
+	rtc_counter = 0;
+	rtc_timeout = 0;
+#endif
+
+	/* Enable RTC module. */
+	rtc_sync();
+	RTC0->CTRL.reg |= RTC_MODE0_CTRL_ENABLE;
+
+	return 0;
+}
+
+void z_clock_set_timeout(s32_t ticks, bool idle)
+{
+	ARG_UNUSED(idle);
+
+#ifdef CONFIG_TICKLESS_KERNEL
+
+	if (ticks < 0 || ticks > MAX_TICKS) {
+		/* Handle K_FOREVER and too big ticks values. */
+		ticks = MAX_TICKS;
+	}
+
+	/* Compute number of RTC cycles until the next timeout. */
+	u32_t count = rtc_count();
+	u32_t timeout = ticks * CYCLES_PER_TICK
+			+ (count - rtc_last) % CYCLES_PER_TICK;
+
+	/* Round to the nearest tick boundary. */
+	timeout = (timeout + CYCLES_PER_TICK - 1) / CYCLES_PER_TICK
+		  * CYCLES_PER_TICK;
+
+	if (timeout < TICK_THRESHOLD) {
+		/* Trigger the interrupt right away if ticks is too low. */
+		NVIC_SetPendingIRQ(DT_RTC_SAM0_IRQ);
+		return;
+	}
+
+	rtc_sync();
+	RTC0->COMP[0].reg = count + timeout;
+
+#else /* !CONFIG_TICKLESS_KERNEL */
+
+	if (ticks < 0) {
+		/* Disable comparator for K_FOREVER and other negative
+		 * values.
+		 */
+		rtc_timeout = rtc_counter;
+		return;
+	}
+
+	if (ticks == 0) {
+		/* Trigger the interrupt right away if ticks is zero. */
+		NVIC_SetPendingIRQ(DT_RTC_SAM0_IRQ);
```

Comment:
Don't do this.  Ticks <= 1 just means "at the next tick".  There's no desire on the part of the kernel to get an interrupt "right now", and in fact that will throw off the timekeeping synchronization and the ticks will get "out of phase" with each other.